### PR TITLE
[automate-2697] stabilize project mgmt tests

### PR DESCRIPTION
**Pull request by bcmdarroch**
### :nut_and_bolt: Description: What code changed, and why?

Stabilized existing tests and marked them unflaky.
Also broke up `project_mgmt` into `project_mgmt`, testing just Projects CRUD, and `project_rule_mgmt`, testing just Rules CRUD.

This we have Cypress coverage for the [Projects GUI section  and most of the Project Rules flow](https://docs.google.com/spreadsheets/d/1mV8dAARhBFxxMZR6gEm1yDXnPqHb9U2mPhZ-zhHLajw/edit#gid=218075547).

What's missing is testing the project update in the UI. I'd like to make that it's own test, since it will involve a lot of moving parts: adding some resources to projects, loading some nodes that we'll write rules for, granting a new user project permissions and then having them log in to verify permissions. I've created a [new tech debt card](https://github.com/chef/automate/issues/3594) for that work.

To run them locally, follow the [e2e Readme](https://github.com/chef/automate/tree/master/e2e#running-cypress-locally)

### :+1: Definition of Done
- [x] `project mgmt` and `project rule mgmt` pass buildkite consistently

### :white_check_mark: Checklist
- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: e2e/cypress/integration/ui/common/global_filter.spec.ts [hunk 1]
```diff
@@ -1,28 +1,30 @@
...
   };
   const proj3: Project = {
-    id: `cypress-project-3-${Cypress.moment().format('MMDDYYhhmm')}`,
+    id: `${cypressPrefix}-proj3-${now}`,
     name: 'Cypress Project 3',
     skip_policies: true
    };
-  const pol_id = `cypress-policy-${Cypress.moment().format('MMDDYYhhmm')}`;
-  const nonAdminUsername = `nonadmin-${Cypress.moment().format('MMDDYYhhmm')}`;
+  const pol_id = `${cypressPrefix}-pol-${now}`;
```
Comment (srenatus) [inline e2e/cypress/integration/ui/common/global_filter.spec.ts:22]: [nit] it's a bit uncommon to use snakecase, I believe
```suggestion
  const polId = `${cypressPrefix}-pol-${now}`;
```

File: e2e/cypress/integration/ui/common/global_filter.spec.ts [hunk 2]
```diff
@@ -58,24 +64,6 @@ describe('global projects filter', () => {
   });
 });
 
-function cleanupProjects(id_token: string): void {
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/common/global_filter.spec.ts:61]: this isn't needed since we have our handy `cy.cleanupIAMObjectsByIDPrefixes` (which also knows how to delete a project's rules before deleting it)

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 3]
```diff
@@ -1,31 +1,33 @@
-import { itFlaky } from '../../../support/constants';
-interface Project {
-  id: string;
-  name: string;
-  type: string;
-}
+import { Project, Rule } from '../../../support/types';
 
 describe('project management', () => {
-  // we increase the default delay to mimic the average human's typing speed
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:9]: decided to drop the `typeDelay` here. It was added as an early attempt to deflake the tests, since sometimes characters were missed when typing. but we've since found better ways to improve stability, mostly around adding expectations to ensure the page has loaded before interacting with elements.

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 4]
```diff
@@ -1,31 +1,33 @@
...
     cy.adminLogin('/settings/projects').then(() => {
       const admin = JSON.parse(<string>localStorage.getItem('chef-automate-user'));
       adminIdToken = admin.id_token;
-      cy.cleanupIAMObjectsByIDPrefixes(cypressPrefix, ['projects', 'policies']);
-      // Set the projects filter to all projects
-      cy.applyProjectsFilter([]);
+      cy.cleanupIAMObjectsByIDPrefixes(cypressPrefix, ['projects']);
+
+      // if there are projects that already exist, set the projects filter to all projects
+      cy.request({
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:22]: might want to turn this into a helper, like `clearProjectsFilter`

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 5]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
-    cy.get('app-create-object-modal chef-button').contains('Cancel').click();
-  });
-
-  it('can open the new project\'s details page', () => {
-    cy.get('[data-cy=project-details]').contains(projectName).click();
-    cy.url().should('include', `/settings/projects/${projectID}`);
-  });
-
-  // something is occasionally resetting the projects filter to (unassigned)
-  itFlaky('can create a rule for the new project', () => {
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:101]: moved these to `project_rule_mgmt`

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 6]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+    cy.get('app-notification.info').contains(`Created project ${projectID}`);
+    cy.get('app-notification.info chef-icon').click();
+
     cy.contains(projectName).should('exist');
     cy.contains(projectID).should('exist');
 
     cy.url().should('include', '/settings/projects');
   });
 
-  it('addPolicies checkbox is always checked on modal open', () => {
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:86]: this is now just an assertion in `can create a project with a custom ID and generate associated policies`

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 7]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+      expect(response.status).to.equal(200);
+    });
+
+    // navigate back to the project list page and get latest project data
+    cy.route('GET', '/apis/iam/v2/projects').as('getProjects');
+    cy.get('.breadcrumb').click();
+    cy.wait('@getProjects');
+
+    cy.get('app-project-list chef-td').contains(customWithoutPolsProjectID).parent()
+      .find('.mat-select-trigger').as('controlMenu');
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:223]: I always had trouble selecting the right element for the control menu in these tests and getting a successful interaction, but it looks like our `mat-select` refactor made it easier to target :)

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 8]
```diff
@@ -252,35 +273,8 @@ describe('project management', () => {
...
       } else {
         cy.get('app-project-list chef-tbody chef-td')
-          .contains(projectID).should('not.exist');
+          .contains(customWithPolsProjectID).should('not.exist');
       }
     });
   });
-
-  // these tests are currently interdependent so mark as flaky until the above isn't
-  itFlaky('can create a project with a custom ID', () => {
```
Comment (bcmdarroch) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:261]: moved this up with the other `create` tests and changed the ID so we wouldn't run into conflicts

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 9]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
-      cy.get('[data-cy=attribute-dropdown]').select(att);
-    });
+    cy.get('[data-cy=create-id]').should('not.be.visible');
+    cy.get('[data-cy=edit-button]').contains('Edit ID').click();
+    cy.get('[data-cy=id-label]').should('not.be.visible');
 
-    cy.get('[data-cy=operator-dropdown]').select('member of');
-    cy.get('[data-cy=operator-dropdown]').select('equals');
-    cy.get('[data-cy=rule-value] input').type('Chef Policy Group Foo');
+    // don't create associated project policies
```
Comment (msorens) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:102]: Helpful to provide a bit more context
```suggestion
    // checkbox is selected by default; uncheck so we do NOT create associated project policies
```

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 4 above)
Comment (blakestier) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:22]: if we're not doing it here, that would be a helpful comment within the code so as we duplicate the work it stands out

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 10]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+    cy.get('app-notification.info').contains(`Created project ${customWithoutPolsProjectID}`);
+    cy.get('app-notification.info chef-icon').click();
+
+    cy.contains(customWithoutPolsProjectID).should('exist');
+
+    // verify no associated policies were generated
+    associatedPolicies.forEach((policy) => {
+      cy.request({
+        auth: { bearer: adminIdToken },
+        url: `/apis/iam/v2/policies/${customWithoutPolsProjectID}-${policy}`,
```
Comment (blakestier) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:121]: this is sort of hard to parse at first glance. I wonder if there's a way to make it clearer? 

maybe `associatedPolicies` -> `appendedPolicyIds`
`policy` -> `policy_id`

and or maybe create a new variable above the request instead of inside it with a new var name? Like `autoCreatedPolicyId` or something? 

🤷

Comment (blakestier) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:121]: maybe i'm hung up on `associated` over `autoGenerated`? 🙃

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 11]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
-    ['Name', 'ID', 'Resource Type', 'Conditions', 'Edits'].forEach((header) => {
-      cy.get('app-project-details chef-th').contains(header);
+    // verify success notification and then dismiss it
+    cy.get('app-notification.info').contains(`Created project ${customWithoutPolsProjectID}`);
+    cy.get('app-notification.info chef-icon').click();
+
+    cy.contains(customWithoutPolsProjectID).should('exist');
+
+    // verify no associated policies were generated
+    associatedPolicies.forEach((policy) => {
```
Comment (msorens) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:118]: This confused me for a moment; would be a bit more clear with this:
```suggestion
    associatedPolicySuffixes.forEach(suffix => {
```

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 12]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+      conditions: [
+        {
+          attribute: 'CHEF_ORGANIZATION',
+          operator: 'EQUALS',
+          values: ['foo']
+        }
+      ]
+    };
+    // add rule via the API to save time,
+    // since we test rules more thoroughly in project_rule_mgmt.spec.ts
```
Comment (blakestier) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:207]: 🎉

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 13]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+
+    cy.get('app-project-list chef-td').contains(customWithoutPolsProjectID).parent()
+      .find('.mat-select-trigger').as('controlMenu');
 
-    cy.get('app-project-details chef-td').contains(ruleID).parent()
-      .find('mat-select').as('controlMenu');
-    // we throw in a should so cypress retries until introspection allows menu to be shown
-    cy.get('@controlMenu').should('be.visible')
+    // we throw in a `should` so cypress retries until introspection allows menu to be shown
+    cy.get('@controlMenu').scrollIntoView().should('be.visible')
```
Comment (blakestier) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:226]: oooooh scroll into vieewwwww

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 14]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
-    cy.get('app-project-details chef-tbody').should('not.exist');
-    cy.get('app-project-details').contains('Edits are pending').should('not.exist');
+    //  close modal
+    cy.get('app-project-list app-message-modal chef-button:first').click();
+
+    // delete rule
+    cy.request({
+      auth: { bearer: adminIdToken },
+      method: 'DELETE',
+      url: `/apis/iam/v2/projects/${customWithoutPolsProjectID}/rules/${rule.id}`
```
Comment (blakestier) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:241]: Should we check for the "Project UPdate" banner?

File: e2e/cypress/integration/ui/settings/project_rule_mgmt.spec.ts [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    cy.get('[data-cy=operator-dropdown]').select('member of');
+    cy.get('[data-cy=operator-dropdown]').select('equals');
+    cy.get('[data-cy=rule-value]').type('Chef Policy Group Foo');
+
+    cy.get('app-project-rules #right-buttons button').contains('Save Rule').click();
+    cy.get('app-project-rules chef-page').should('not.be.visible');
+
+    cy.url().should('include', `/settings/projects/${project.id}`);
+    cy.get('app-project-details chef-td').contains(ruleID);
+    cy.get('app-pending-edits-bar').contains('Project edits pending');
```
Comment (blakestier) [inline e2e/cypress/integration/ui/settings/project_rule_mgmt.spec.ts:75]: is this the project update banner?

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts [hunk 16]
```diff
@@ -75,171 +77,190 @@ describe('project management', () => {
...
+      method: 'POST',
+      url: `/apis/iam/v2/projects/${customWithoutPolsProjectID}/rules`,
+      body: rule
+    }).then((response) => {
+      expect(response.status).to.equal(200);
+    });
+
+    // navigate back to the project list page and get latest project data
+    cy.route('GET', '/apis/iam/v2/projects').as('getProjects');
+    cy.get('.breadcrumb').click();
```
Comment (msorens) [inline e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:219]: 1. Did the CreateRule change the current page? I thought we were on the ListProjects page at the start of this test, but if you are clicking on a breadcrumb, then apparently not...?
2. Assuming on the project detail page, then, and the breadcrumb click takes us back to the ListProjects page, then why is the explicit projects fetch needed? Isn't that done by the breadcrumb click navigation?

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 14 above)
Comment (blakestier) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:241]: (I know we don't want to test the functionality of the banner here, however, and we could remove when we add the fuller update tests)

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 4 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:22]: Looks like this is the only instance we're using it, so I'll leave this as is and updatethe comment

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 10 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:121]: went with a combo of yours and @msorens' suggestions

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 16 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:219]: The `cy.route` line isn't actually fetching the data, it specifies what API call Cypress should wait for on L220 (`cy.wait('@getProjects');`)

So we return to the list page with the breadcrumb, wait for the ListProjects call to come back, then proceed with the rest of the test.

That being said, I might rearrange this a bit.

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 13 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:226]: yeah pretty slick 😎 might need to use that in some of our other tests

File: e2e/cypress/integration/ui/settings/project_rule_mgmt.spec.ts (hunk 15 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_rule_mgmt.spec.ts:75]: Yeah, I can add a comment

File: e2e/cypress/integration/ui/settings/project_mgmt.spec.ts (hunk 14 above)
Comment (bcmdarroch) [reply e2e/cypress/integration/ui/settings/project_mgmt.spec.ts:241]: sure
